presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/AddProjectAboveCrossJoin.java

```diff
@@ -0,0 +1,79 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.facebook.presto.sql.planner.iterative.rule;
+
+import com.facebook.presto.matching.Captures;
+import com.facebook.presto.matching.Pattern;
+import com.facebook.presto.spi.plan.ProjectNode;
+import com.facebook.presto.spi.relation.VariableReferenceExpression;
+import com.facebook.presto.sql.planner.iterative.Rule;
+import com.facebook.presto.sql.planner.plan.JoinNode;
+import com.google.common.collect.ImmutableSet;
+
+import java.util.ArrayList;
+import java.util.Set;
+
+import static com.facebook.presto.spi.plan.ProjectNode.Locality.LOCAL;
+import static com.facebook.presto.sql.planner.plan.AssignmentUtils.identityAssignments;
+import static com.facebook.presto.sql.planner.plan.Patterns.join;
+
+/**
+ * Cross joins don't support output symbol selection, so for any such node where (input variables != output variables),
```

Comment:
If the change in predicate pushdown is to to produce cross join node which violates this assertion, I strongly suggest to put the logic here into predicate pushdown. As my understanding is that, the current approach will produce invalid intermediate plans with cross join node violating the assertion, and relies on this optimizer to fix it at the end, which is error prone.

Reply:
> [This](https://github.com/prestodb/presto/blob/master/presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/PredicatePushDown.java#L605-L607) check, in particular, adds a Project over a Join node even when the JoinNode is logically capable of pruning it's input.

Then I think what @rschlussel suggested below is a better option.